File: package/main/src/Validate/object/optional.ts
import {
  attachStandard,
  type StandardSchemaV1,
} from "@/Validate/standardSchema";

export interface UndefinedReturn {
  validate: boolean;
  message: string;
  type: "undefined";
}

/**
 * Optional validator augmented with a reference to the wrapped validator,
 * used by `required()` to unwrap optional layers when rebuilding a shape
 * @template T - The type of value the wrapped validator expects
 * @template R - The return type of the wrapped validator (preserved so the inner type tag flows through)
 */
export type OptionalValidator<
  T,
  R extends { type: unknown; message: string; validate: boolean } = {
    type: unknown;
    message: string;
    validate: boolean;
  },
> = ((value?: T) => R | UndefinedReturn) & {
  inner: (value: T) => R;
  isOptional: true;
};

/**
 * Wraps a validator to accept undefined values
 * @template T - The type of value the wrapped validator expects
 * @template R - The return type of the wrapped validator (preserved so the inner type tag flows through)
 * @param {Function} validator - Validator function to make optional
 * @returns {OptionalValidator<T, R>} - Validator that passes for undefined or delegates to the wrapped validator
 */
export const optional = <
  T,
  R extends { type: unknown; message: string; validate: boolean },
>(
  validator: (value: T) => R,
): OptionalValidator<T, R> & StandardSchemaV1<T | undefined, T | undefined> => {
  const optionalValidator = ((value?: T): R | UndefinedReturn => {
    if (value === undefined) {
      return {
        validate: true,
        message: "",
        type: "undefined",
      };
    }
    return validator(value);
  }) as OptionalValidator<T, R>;

  optionalValidator.inner = validator;
  optionalValidator.isOptional = true;

  return attachStandard<T | undefined, T | undefined, typeof optionalValidator>(
    optionalValidator,
  );
};

File: package/umt_go/src/validate/optional.go
package validate

// Optional wraps a property validator so it also accepts an absent (undefined)
// value, mirroring the TypeScript `optional()` validator
// (Validate/object/optional.ts) and the Rust `umt_partial` per-property wrapper
// (validate/object/partial.rs), which both let an absent value pass while
// delegating a present value to the inner validator.
//
// The returned PropertyValidator treats a nil incoming value as JavaScript
// `undefined` (the Rust `None`): it is reported valid with an empty message and
// the literal "undefined" tag through the Type field, mirroring the TypeScript
// `UndefinedReturn`. A present (non-nil) value is delegated to the wrapped
// property validator and its result flows through unchanged.
//
// The result carries IsOptional == true and Inner pointing at the wrapped
// validator, reproducing the `isOptional` flag and `inner` reference the
// TypeScript `optional()` attaches to its function so Required can unwrap the
// layer and Partial can stay idempotent (Go cannot attach properties to a bare
// function, so this metadata travels on the PropertyValidator struct).
func Optional(validator PropertyValidator) PropertyValidator {
	inner := validator
	return PropertyValidator{
		Validate: func(value any) ValidateCoreReturnType[any] {
			if value == nil {
				return ValidateCoreReturnType[any]{
					Validate: true,
					Message:  "",
					Type:     "undefined",
				}
			}
			return inner.Validate(value)
		},
		IsOptional: true,
		Inner:      &inner,
	}
}
